### PR TITLE
Add support for starting suffix numbers

### PR DESCRIPTION
**Pull request by andrewbaptist**
This commit now allows split to pass split/numeric.sh

### CODE REVIEW

**Review by tertsdiepraam — COMMENTED** [review]
Very cool! One nit about the for loop. I agree that it might not be slow, but it still feels wrong. It's too bad that `Number` does not already have some convenient method of this.

File: src/uu/split/src/filenames.rs [hunk 1]
```diff
@@ -161,51 +173,86 @@ mod tests {
...
         assert_eq!(it.next().unwrap(), "chunk_02.txt");
 
-        let mut it = FilenameIterator::new("chunk_", ".txt", 0, SuffixType::Decimal);
+        let mut it = FilenameIterator::new("chunk_", ".txt", 0, SuffixType::Decimal, 0).unwrap();
         assert_eq!(it.nth(10 * 9 - 1).unwrap(), "chunk_89.txt");
         assert_eq!(it.next().unwrap(), "chunk_9000.txt");
         assert_eq!(it.next().unwrap(), "chunk_9001.txt");
     }
+
+    #[test]
```
Comment (jfinkels) [inline src/uu/split/src/filenames.rs:224]: Good tests. Could you also write one in the `tests/by-util/test_split.rs`?

File: src/uu/split/src/filenames.rs [hunk 2]
```diff
@@ -119,19 +120,30 @@ impl<'a> FilenameIterator<'a> {
...
+    ) -> UResult<FilenameIterator<'a>> {
         let radix = suffix_type.radix();
-        let number = if suffix_length == 0 {
+        let mut number = if suffix_length == 0 {
             Number::DynamicWidth(DynamicWidthNumber::new(radix))
         } else {
             Number::FixedWidth(FixedWidthNumber::new(radix, suffix_length))
         };
-        FilenameIterator {
+        // Not very efficient, but typically the suffixes are small.
```
[the remark this answers is not in the document]
Comment (jfinkels) [reply src/uu/split/src/filenames.rs:131]: another alternative is to add a function like `DynamicWidthNumber::starting_at(radix, suffix_start)` (and `FixedWidthNumber::starting_at(radix, suffix_length, suffix_start)`) to encapsulate the functionality in this loop

**Review by jfinkels — COMMENTED** [review]
Great!

File: src/uu/split/src/filenames.rs (hunk 1 above)
Comment (andrewbaptist) [reply src/uu/split/src/filenames.rs:222]: Done.

File: src/uu/split/src/filenames.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewbaptist) [reply src/uu/split/src/filenames.rs]: OK - a little more refactoring than expected, but the numeric class is much nicer now.

File: src/uu/split/src/number.rs [hunk 3]
```diff
@@ -467,17 +434,17 @@ mod tests {
...
-        assert_eq!(format!("{}", num(10 * 9)), "9000");
-        assert_eq!(format!("{}", num(10 * 9 + 1)), "9001");
-        assert_eq!(format!("{}", num(10 * 99 - 1)), "9899");
-        assert_eq!(format!("{}", num(10 * 99)), "990000");
-        assert_eq!(format!("{}", num(10 * 99 + 1)), "990001");
+        assert_eq!(format!("{}", num(10 * 9)), "90");
+        assert_eq!(format!("{}", num(10 * 9 + 1)), "91");
+        assert_eq!(format!("{}", num(10 * 99 - 1)), "989");
+        assert_eq!(format!("{}", num(10 * 99)), "990");
+        assert_eq!(format!("{}", num(10 * 99 + 1)), "991");
```
Comment (tertsdiepraam) [inline src/uu/split/src/number.rs:441]: This was actually very intentional behaviour (and matched GNU), implemented in https://github.com/uutils/coreutils/pull/2859. Consider what happens to the filename sorting in your implementation:
```
some_file_98.txt
some_file_99.txt
some_file_100.txt
# gets sorted into this (due to lexical order of strings):
some_file_100.txt
some_file_98.txt
some_file_99.txt
```
The current algorithm, however, keeps `9`s at the start to ensure being sorted after earlier files.

Comment (tertsdiepraam) [reply src/uu/split/src/number.rs:441]: Okay so looking at this again I made an important discovery for this PR. With the `--numeric-suffix` option, the dynamic number is never used. It's always fixed width with a default width of 2. Here is an example:
```shell
$ split Cargo.lock Foo_ -l 1 --numeric-suffixes=80
split: output file suffixes exhausted
[ creates Foo_80 to Foo_99 ]
$ split Cargo.lock Foo_ -l 1 --numeric-suffixes=80 --suffix-length=4 
[ creates Foo_80 to Foo_3376 ]
$ split Cargo.lock Foo_ -l 1 -d          # warning creates a lot of files
[ creates Foo_00 up to Foo_992124 (which is 3307 files) ]
```

**Review by tertsdiepraam — COMMENTED** [review]
Okay, turns out `--suffix-length` always uses fixed width. So we can simplify this a bit, to only have the initial number for fixed width.

File: src/uu/split/src/number.rs (hunk 3 above)
Comment (andrewbaptist) [reply src/uu/split/src/number.rs:440]: OK - I'll take another pass at this later this week. It makes more sense now how it works, but it is a little crazy!

**Comment by sylvestre** [comment]
sorry but needs a small rebase

**Comment by sylvestre** [comment]
fails on:
```

---- test_split::test_numeric_dynamic_suffix_length stderr ----
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', tests/common/util.rs:542:37
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    test_split::test_alphabetic_dynamic_suffix_length
    test_split::test_hex_dynamic_suffix_length
    test_split::test_numeric_dynamic_suffix_length

```

**Comment by andrewbaptist** [comment]
I plan to finish this today. I have a easy way to compute the names without counting.

**Comment by andrewbaptist** [comment]
OK - I think it is ready for review again. I haven't compared the output with the output from gnu split completely yet, but I think the current behavior is much better. The "initial" value for dynamic is a little confusing and could possibly be improved upon.

**Comment by andrewbaptist** [comment]
Let me know if you need anything more changes on this PR.

**Comment by sylvestre** [comment]
terrific:
`Warning: Congrats! The gnu test tests/split/numeric is no longer failing!`

**Pull request merged**